doc/decisions/constructor_functions_examples.md

```diff
@@ -0,0 +1,105 @@
+
+```c
+Key * k1 = elektraKeyCreate (NS_USER, elektraKeyNameEmpty);
+Key * k2 = elektraKeyCreate (NS_USER, elektraKeyName("foo", "bar"));
+Key * k3 = elektraKeyCreateS (NS_USER, elektraKeyName("foo", "foo"), "value");
+Key * k4 = elektraKeyCreateV (NS_USER, elektraKeyName("foo", "foo", "bar"), "value", 3);
+Key * k5 = elektraKeyCreateSM (NS_USER, elektraKeyName("foo", "baz"), "value", elektraKeyCreateM(elektraKeyName("type"), "string"), elektraKeyCreateM(elektraKeyName("other"), "aaa"));
+Key * k6 = elektraKeyCreateVM (NS_USER, elektraKeyName("foo", "qux"), "value", 4, elektraKeyCreateM(elektraKeyName("type"), "string"));
+```
+
+The definitions of the macros are complicated, but this would expand to something like:
+
+```c
+Key * k1 = elektraKeyBuild(NS_USER, ((struct KeyName){ NULL, 0 }), NULL, 0, NULL);
+Key * k2 = elektraKeyBuild(NS_USER, ((struct KeyName){"foo" "\0" "bar", 0 + strlen("bar") + 1 + strlen("foo" "\0" "bar")}), NULL, 0, NULL);
+Key * k3 = elektraKeyBuild(NS_USER, ((struct KeyName){"foo" "\0" "foo", 0 + strlen("foo") + 1 + strlen("foo" "\0" "foo")}), "value", strlen("value") + 1, NULL);
+Key * k4 = elektraKeyBuild(NS_USER, ((struct KeyName){"foo" "\0" "foo" "\0" "bar", 0 + strlen("bar") + 1 + strlen("foo" "\0" "bar") + 1 + strlen("foo" "\0" "foo" "\0" "bar")}), "value", 3, NULL);
+Key * k5 = elektraKeyBuild(NS_USER, ((struct KeyName){"foo" "\0" "baz", 0 + strlen("baz") + 1 + strlen("foo" "\0" "baz")}), "value", strlen("value") + 1, ksNew(
+                elektraKeyBuild(NS_META, ((struct KeyName){"type", 0 + strlen("type")}), "string", strlen("string") + 1, NULL),
+                elektraKeyBuild(NS_META, ((struct KeyName){"other", 0 + strlen("other")}), "aaa", strlen("aaa") + 1, NULL)
+        ));
+Key * k6 = elektraKeyBuild(NS_USER, ((struct KeyName){"foo" "\0" "qux", 0 + strlen("qux") + 1 + strlen("foo" "\0" "qux")}), "value", 4, ksNew(
+                elektraKeyBuild(NS_META, ((struct KeyName){"type", 0 + strlen("type")}), "string", strlen("string") + 1, NULL)
+        ));
+```
+
+The `elektraKeyName` macro only works for literal strings, so if the name is not known at compile time, you'd need to use `elektraKeyBuild` directly:
+
+```c
+struct KeyName {
+    const char * name; // unescaped name, contains '\0' chars
+    size_t size;
+};
+
+Key * elektraKeyBuild(elektraNamespace ns, struct KeyName name, const void * value, size_t vSize, const KeySet * meta);
+```
+
+However, here we can also add a secondary function `elektraKeyBuildFromParts`, which would build the name from parts with multiple `memcpy` (something that cannot be avoided with non-literal strings):
+
+```c
+Key * elektraKeyBuildFromParts(elektraNamespace ns, const char ** nameParts, const void * value, size_t vSize, const KeySet * meta);
+```
+
+This could be use with non-literal strings like so:
+
+```c
+elektraKeyBuildFromParts(NS_USER, ((const char*[]) { "foo", bar, "baz" }), "value", sizeof("value"), NULL);
```

Comment:
Just noticed, this doesn't work. We need a `NULL` terminator (or a `partCount`):

```suggestion
elektraKeyBuildFromParts(NS_USER, ((const char*[]) { "foo", bar, "baz", NULL }), "value", sizeof("value"), NULL);
```